src/python/pants/engine/legacy/graph.py

```diff
@@ -360,24 +361,39 @@ def transitive_dependents_of_addresses(self, addresses):
         return result
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass
 class HydratedTarget:
     """A wrapper for a fully hydrated TargetAdaptor object.
 
+    Why have this type if all information is stored in the underlying TargetAdaptor? We need it for
+    the type-driven graph to work properly. A HydratedTarget is a generic wrapper around different
+    target classes, whereas a TargetAdaptor has different types like PythonTestsAdaptor and
+    PythonLibraryAdaptor. We need to be able to work with targets both generically and depending
+    on their target type (i.e. unions).
+
     Transitive graph walks collect ordered sets of TransitiveHydratedTargets which involve a huge
-    amount of hashing: we implement eq/hash via direct usage of an Address field to speed that up.
+    amount of hashing: we implement eq/hash via a private Address field to speed that up.
     """
 
-    address: Address
     adaptor: TargetAdaptor
-    dependencies: Tuple[Address, ...]
 
-    @property
-    def addresses(self) -> Tuple:
-        return self.dependencies
+    def __init__(self, adaptor: TargetAdaptor) -> None:
+        self.adaptor = adaptor
+        # This field is set for efficient lookup of the address in `__hash__` and `__eq__` during
+        # graph walks. Directly accessing this field, rather than using `adaptor.address` is about
+        # 10x faster. However, rules should still use `adaptor.address` for clarity and because
+        # they do not need to access the address as frequently, so it would be an over-optimization
+        # (we're talking about 0.000_000_1 vs. 0.000_001 seconds here).
+        self._address = adaptor.address
 
     def __hash__(self):
-        return hash(self.address)
+        return hash(self._address)
+
+    def __eq__(self, other):
+        if not isinstance(other, HydratedTarget):
+            return NotImplemented
+        return self._address == other._address
```

Comment:
It was an oversight to not implement this before. Whenever you override `__hash__`, you should override `__eq__` so that the two correspond.

Adding this sped up the graph construction in `build-support/bin/mypy.py` from 4.7 to 4.3 seconds. (The final result was 3.8.)

--

Btw, to test performance, add this diff:

```diff
diff --git a/src/python/pants/bin/pants_loader.py b/src/python/pants/bin/pants_loader.py
index e71aa2558..800929e5e 100644
--- a/src/python/pants/bin/pants_loader.py
+++ b/src/python/pants/bin/pants_loader.py
@@ -87,7 +87,9 @@ class PantsLoader:
         cls.setup_warnings()
         cls.ensure_locale()
         entrypoint = cls.determine_entrypoint(cls.ENTRYPOINT_ENV_VAR, cls.DEFAULT_ENTRYPOINT)
-        cls.load_and_execute(entrypoint)
+        from pants.util.contextutil import maybe_profiled
+        with maybe_profiled("dedupe_ht.prof"):
+            cls.load_and_execute(entrypoint)
 
 
 def main():
```

Then `pip install snakeviz`, `build-support/bin/mypy.py`, and `snakeviz dedupe_ht.prof`.